File: lib/std/time/calendar-inline.js
/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

function _local_get_timezone() {
  return null; // not possible in javascript, it's a global :-(
}

var dtformat = undefined;

function _local_get_timezone_abbrv(d) {
  // try to get the timezone abbreviation; expensive to initialize so do that once only;
  // even with static initialization, getTimezoneOffset and format are pretty slow...
  if (typeof dtformat === "undefined") {
    try {
      dtformat = Intl.DateTimeFormat("en-US",{timeZoneName:"short"});
    }
    catch(exn) {
      dtformat = null;
    }
  }
  return (dtformat == null ? "" : dtformat.format(d).replace(/^.*\s([\w\+\-:]{3,6})$/,"$1"));
}

function _local_utc_delta(tz,i) {
  var d = new Date(i*1000);
  
  // if we have a datetime format, we get the timezone abbreviation
  var abbrv = _local_get_timezone_abbrv(d);
  
  // Return the timezone offset (switch sign!) and the timezone abbreviation
  return $std_core._tuple2_( (d.getTimezoneOffset())*-60, abbrv );
}

File: lib/std/time/calendar-inline.cs
/*---------------------------------------------------------------------------
  Copyright 2016-2017 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

static class _Calendar {
  private static long unixEpochTicks = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc).Ticks;

  public static __std_core._Tuple2_<double,string> LocalUtcDelta( object tzi, double unixSeconds ) {
    TimeZoneInfo tz = tzi as TimeZoneInfo;
    if (tz==null) tz = TimeZoneInfo.Local;
    
    double secs     = Math.Floor(unixSeconds);
    long unixTicks  = (long)secs * TimeSpan.TicksPerSecond + (long)((unixSeconds - secs)*(double)TimeSpan.TicksPerSecond);
    DateTime t      = new DateTime(unixEpochTicks + unixTicks, DateTimeKind.Utc);
    double ofs      = tz.GetUtcOffset( t ).TotalSeconds;
    string abbrv    = (tz.IsDaylightSavingTime(t) ? tz.DaylightName : tz.StandardName); 
    return new __std_core._Tuple2_<double,string>( ofs, abbrv );
  }

  public static TimeZoneInfo LocalTimeZone() {
    return TimeZoneInfo.Local;
  }
}